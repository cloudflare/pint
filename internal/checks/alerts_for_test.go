package checks_test

import (
	"testing"

	"github.com/cloudflare/pint/internal/checks"
	"github.com/cloudflare/pint/internal/output"
	"github.com/cloudflare/pint/internal/parser"
	"github.com/cloudflare/pint/internal/promapi"
)

func newAlertsForCheck(_ *promapi.FailoverGroup) checks.RuleChecker {
	return checks.NewAlertsForCheck()
}

func TestAlertsForCheck(t *testing.T) {
	testCases := []checkTest{
		{
			description: "ignores rules with syntax errors",
			content:     "- alert: foo\n  expr: sum(foo) without(\n",
			checker:     newAlertsForCheck,
			prometheus:  noProm,
			problems:    noProblems,
		},
		{
			description: "ignores recording rules",
			content:     "- record: foo\n  expr: sum(foo) without(job)\n",
			checker:     newAlertsForCheck,
			prometheus:  noProm,
			problems:    noProblems,
		},
		{
			description: "invalid for value",
			content:     "- alert: foo\n  expr: foo\n  for: abc\n",
			checker:     newAlertsForCheck,
			prometheus:  noProm,
			problems: func(_ string) []checks.Problem {
				return []checks.Problem{
					{
						Lines: parser.LineRange{
							First: 3,
							Last:  3,
						},
						Reporter: "alerts/for",
						Summary:  `invalid duration`,
						Details:  checks.AlertForCheckDurationHelp,
						Severity: checks.Bug,
						Diagnostics: []output.Diagnostic{
							{
								Line:        3,
								FirstColumn: 8,
								LastColumn:  10,
								Message:     `not a valid duration string: "abc"`,
							},
						},
					},
				}
			},
		},
		{
			description: "negative for value",
			content:     "- alert: foo\n  expr: foo\n  for: -5m\n",
			checker:     newAlertsForCheck,
			prometheus:  noProm,
			problems: func(_ string) []checks.Problem {
				return []checks.Problem{
					{
						Lines: parser.LineRange{
							First: 3,
							Last:  3,
						},
						Reporter: "alerts/for",
						Summary:  `invalid duration`,
						Details:  checks.AlertForCheckDurationHelp,
						Severity: checks.Bug,
						Diagnostics: []output.Diagnostic{
							{
								Line:        3,
								FirstColumn: 8,
								LastColumn:  10,
								Message:     `not a valid duration string: "-5m"`,
							},
						},
					},
				}
			},
		},
		{
			description: "default for value",
			content:     "- alert: foo\n  expr: foo\n  for: 0h\n",
			checker:     newAlertsForCheck,
			prometheus:  noProm,
			problems: func(_ string) []checks.Problem {
				return []checks.Problem{
					{
						Lines: parser.LineRange{
							First: 3,
							Last:  3,
						},
						Reporter: "alerts/for",
						Summary:  "redundant field with default value",
						Severity: checks.Information,
						Diagnostics: []output.Diagnostic{
							{
								Line:        3,
								FirstColumn: 8,
								LastColumn:  9,
								Message:     "`0h` is the default value of `for`, this line is unnecessary.",
							},
						},
					},
				}
			},
		},
		{
			description: "invalid keep_firing_for value",
			content:     "- alert: foo\n  expr: foo\n  keep_firing_for: abc\n",
			checker:     newAlertsForCheck,
			prometheus:  noProm,
			problems: func(_ string) []checks.Problem {
				return []checks.Problem{
					{
						Lines: parser.LineRange{
							First: 3,
							Last:  3,
						},
						Reporter: "alerts/for",
						Summary:  `invalid duration`,
						Details:  checks.AlertForCheckDurationHelp,
						Severity: checks.Bug,
						Diagnostics: []output.Diagnostic{
							{
								Line:        3,
								FirstColumn: 20,
								LastColumn:  22,
								Message:     `not a valid duration string: "abc"`,
							},
						},
					},
				}
			},
		},
		{
			description: "negative keep_firing_for value",
			content:     "- alert: foo\n  expr: foo\n  keep_firing_for: -5m\n",
			checker:     newAlertsForCheck,
			prometheus:  noProm,
			problems: func(_ string) []checks.Problem {
				return []checks.Problem{
					{
						Lines: parser.LineRange{
							First: 3,
							Last:  3,
						},
						Reporter: "alerts/for",
						Summary:  `invalid duration`,
						Details:  checks.AlertForCheckDurationHelp,
						Severity: checks.Bug,
						Diagnostics: []output.Diagnostic{
							{
								Line:        3,
								FirstColumn: 20,
								LastColumn:  22,
								Message:     `not a valid duration string: "-5m"`,
							},
						},
					},
				}
			},
		},
		{
			description: "default for value",
			content:     "- alert: foo\n  expr: foo\n  keep_firing_for: 0h\n",
			checker:     newAlertsForCheck,
			prometheus:  noProm,
			problems: func(_ string) []checks.Problem {
				return []checks.Problem{
					{
						Lines: parser.LineRange{
							First: 3,
							Last:  3,
						},
						Reporter: "alerts/for",
						Summary:  "redundant field with default value",
						Severity: checks.Information,
						Diagnostics: []output.Diagnostic{
							{
								Line:        3,
								FirstColumn: 20,
								LastColumn:  21,
								Message:     "`0h` is the default value of `keep_firing_for`, this line is unnecessary.",
							},
						},
					},
				}
			},
		},
	}
	runTests(t, testCases)
}
