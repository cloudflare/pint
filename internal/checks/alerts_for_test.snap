
[TestAlertsForCheck/default_for_value - 1]
- description: default for value
  content: |
    - alert: foo
      expr: foo
      for: 0h
  output: |
    3 |   for: 0h
               ^^ `0h` is the default value of `for`, this line is unnecessary.
  problem:
    reporter: alerts/for
    summary: redundant field with default value
    details: ""
    diagnostics:
        - message: '`0h` is the default value of `for`, this line is unnecessary.'
          pos:
            - line: 3
              firstcolumn: 8
              lastcolumn: 9
          firstcolumn: 1
          lastcolumn: 2
          kind: 0
    lines:
        first: 3
        last: 3
    severity: 0
    anchor: 0

---

[TestAlertsForCheck/default_for_value#01 - 1]
- description: default for value
  content: |
    - alert: foo
      expr: foo
      keep_firing_for: 0h
  output: |
    3 |   keep_firing_for: 0h
                           ^^ `0h` is the default value of `keep_firing_for`, this line is unnecessary.
  problem:
    reporter: alerts/for
    summary: redundant field with default value
    details: ""
    diagnostics:
        - message: '`0h` is the default value of `keep_firing_for`, this line is unnecessary.'
          pos:
            - line: 3
              firstcolumn: 20
              lastcolumn: 21
          firstcolumn: 1
          lastcolumn: 2
          kind: 0
    lines:
        first: 3
        last: 3
    severity: 0
    anchor: 0

---

[TestAlertsForCheck/ignores_recording_rules - 1]
[]

---

[TestAlertsForCheck/ignores_rules_with_syntax_errors - 1]
[]

---

[TestAlertsForCheck/invalid_for_value - 1]
- description: invalid for value
  content: |
    - alert: foo
      expr: foo
      for: abc
  output: |
    3 |   for: abc
               ^^^ not a valid duration string: "abc"
  problem:
    reporter: alerts/for
    summary: invalid duration
    details: Supported time durations are documented [here](https://prometheus.io/docs/prometheus/latest/querying/basics/#time-durations).
    diagnostics:
        - message: 'not a valid duration string: "abc"'
          pos:
            - line: 3
              firstcolumn: 8
              lastcolumn: 10
          firstcolumn: 1
          lastcolumn: 3
          kind: 0
    lines:
        first: 3
        last: 3
    severity: 2
    anchor: 0

---

[TestAlertsForCheck/invalid_keep_firing_for_value - 1]
- description: invalid keep_firing_for value
  content: |
    - alert: foo
      expr: foo
      keep_firing_for: abc
  output: |
    3 |   keep_firing_for: abc
                           ^^^ not a valid duration string: "abc"
  problem:
    reporter: alerts/for
    summary: invalid duration
    details: Supported time durations are documented [here](https://prometheus.io/docs/prometheus/latest/querying/basics/#time-durations).
    diagnostics:
        - message: 'not a valid duration string: "abc"'
          pos:
            - line: 3
              firstcolumn: 20
              lastcolumn: 22
          firstcolumn: 1
          lastcolumn: 3
          kind: 0
    lines:
        first: 3
        last: 3
    severity: 2
    anchor: 0

---

[TestAlertsForCheck/negative_for_value - 1]
- description: negative for value
  content: |
    - alert: foo
      expr: foo
      for: -5m
  output: |
    3 |   for: -5m
               ^^^ not a valid duration string: "-5m"
  problem:
    reporter: alerts/for
    summary: invalid duration
    details: Supported time durations are documented [here](https://prometheus.io/docs/prometheus/latest/querying/basics/#time-durations).
    diagnostics:
        - message: 'not a valid duration string: "-5m"'
          pos:
            - line: 3
              firstcolumn: 8
              lastcolumn: 10
          firstcolumn: 1
          lastcolumn: 3
          kind: 0
    lines:
        first: 3
        last: 3
    severity: 2
    anchor: 0

---

[TestAlertsForCheck/negative_keep_firing_for_value - 1]
- description: negative keep_firing_for value
  content: |
    - alert: foo
      expr: foo
      keep_firing_for: -5m
  output: |
    3 |   keep_firing_for: -5m
                           ^^^ not a valid duration string: "-5m"
  problem:
    reporter: alerts/for
    summary: invalid duration
    details: Supported time durations are documented [here](https://prometheus.io/docs/prometheus/latest/querying/basics/#time-durations).
    diagnostics:
        - message: 'not a valid duration string: "-5m"'
          pos:
            - line: 3
              firstcolumn: 20
              lastcolumn: 22
          firstcolumn: 1
          lastcolumn: 3
          kind: 0
    lines:
        first: 3
        last: 3
    severity: 2
    anchor: 0

---
